### PR TITLE
Refactor the app to allow choosing between next minor and major update

### PR DESCRIPTION
**Pull request by VicDeo**
#### Core upgrade (this part depends on https://github.com/owncloud/core/pull/32491 that is still WIP)
- major (10 to 11 and like this) still DOES  upgrade apps  from 1.0 to 2.0 (if available)
- minor (10 to 10.1 and like this) does NOT apps upgrade from 1.0 to 2.0

#### CLI
- `market:upgrade` still DOES upgrade 1.0 to 1.1
- `market:upgrade` does NOT upgrade 1.0 to 2.0
- `market:upgrade --major` DOES upgrade 1.0 to 2.0
- `market:install` does NOT upgrade 1.0 to 2.0 if the app is already installed
- `market:install` still DOES upgrade 1.0 to 1.1 if the app is already installed

#### Market app Web UI
- Shows both 1.1 and 2.0 for 1.0, 1.1 is default

#### Market app cron job
- Generates notifications both for 1.1 and 2.0 

#### Updates tab
newupdatespage

#### App Page
apppage

### TODO
- [x] add `--major` option  to CLI `market:upgrade` command
- [x] do not allow update to major releases with CLI `market:install` command
- [x] switch to major version of the app only in case there is a core major update is happening
    (Depends on https://github.com/owncloud/core/pull/32491)
- [x] Update app UI to show both minor and major updates
- [x] Allow update both to major and minor versions from UI

0.3.5 to 0.3.5
```
deo@jah-mobile:> php occ market:upgrade --local ../customgroups.tar.gz 
customgroups: ../customgroups.tar.gz has the same or older version of the app
```
0.3.5 to 1.3.5 (no `--major`)
```
deo@jah-mobile:> php occ market:upgrade --local ../customgroups.tar.gz 
customgroups: ../customgroups.tar.gz has a different major version, try with --major option
```

0.3.5 to 1.3.5 (with `--major`)
```
deo@jah-mobile:> php occ market:upgrade --local ../customgroups.tar.gz --major
customgroups: Installing new version from ../customgroups.tar.gz
```

###  after the merge
- [ ] Documentation needs to be updated

### CODE REVIEW

File: lib/MarketService.php [hunk 1]
```diff
@@ -140,15 +140,22 @@ public function readAppPackage($path){
 		return $this->appManager->readAppPackage($path);
 	}
 
-	private function downloadPackage($appId) {
+	private function downloadPackage($appId, $isMajor, $currentVersion = '0.0.0.0') {
```
Comment (PVince81) [inline lib/MarketService.php]: call it `$allowMajorUpdate` ?

**Comment by PVince81** [comment]
we'll also need the `--major` for `occ market:update` which I guess will use the same methods ?

**Comment by VicDeo** [comment]
Build is failing due to a version bump in core:
>App "Market" cannot be installed because the following dependencies are not  
   fulfilled: ownCloud 10.1 or lower is required.

**Comment by VicDeo** [comment]
given that I am not an
 UI expert I appreciate if someone explain how this page should look like if we have both minor an major updates...

updatesui
apppageui

File: lib/MarketService.php (hunk 1 above)
Comment (PVince81) [reply lib/MarketService.php]: it is not clear to me whether $isMajor says that we're dealing with a core major update, or if we're requesting `downloadPackage` to only download package compatible with a major update of $currentVersion

**Comment by PVince81** [comment]
@felixheidecke any thoughts ?

We could split the "Update info" column into three:
- Installed version
- Minor update target
- Major update target

Might then need two buttons, or make the update target clickable to select which one to update to.

Or a dropdown to select the version ?

**Comment by PVince81** [comment]
and now I realize that whatever the UX will be it will have an impact on the info that the controller endpoints must return...

**Comment by VicDeo** [comment]
@PVince81 it has even more impact than you can imagine ;)
market:upgrade --list 
updated to the following output
```
deo@jah-mobile:~> php occ market:upgrade --list
admin_audit : minor:0.9.0, major:1.0.1
enterprise_key : minor:0.1.4
onlyoffice : minor:1.4.0, major:2.0.3
theme-enterprise : major:2.0.2
windows_network_drive : minor:0.7.1
```


I managed to pass both versions to frontend and display them
screenshot_20181029_000849

Imo for two candidates the right side should mutate into the HTML select element

**Comment by PVince81** [comment]
@VicDeo go for it :-D

Hint: https://getuikit.com/v2/docs/dropdown.html

**Comment by VicDeo** [comment]
@PVince81 another questionable thing is `market:install` command. It updates the app if it is already installed. 
As `market:install --major` looks pretty non-sense IMHO it should refuse to operate on major versions. what do you think?

**Comment by PVince81** [comment]
@VicDeo if `market:install` is only doing installations and not updates, then the switch `--major` doesn't make sense as there is no local version. Installing an app should always return the latest version I guess.

Is there a possibility to specify a version number to install like with Linux package managers ?

**Comment by VicDeo** [comment]
@PVince81 >Is there a possibility to specify a version number to install like with Linux package managers ?

no, it installs the most recent, or update to the most recent if the app is already installed.

**Comment by PVince81** [comment]
@VicDeo how long do you think you still need to finish this PR ?

**Comment by VicDeo** [comment]
@PVince81 it's hard to say as I still suck in the frontend part :(

**Comment by PVince81** [comment]
@VicDeo you can get some advice from @felixheidecke in the chat

**Comment by VicDeo** [comment]
@PVince81 @felixheidecke  I updated the PR description with new look for `Updates` tab

I appreciate the advice on the app details tab given that both releases could have missing dependencies section here...
appdetails

**Comment by VicDeo** [comment]
What would be correct to use from the point of UI here, radiobutton?

**Comment by felixheidecke** [comment]
> What would be correct to use from the point of UI here, radiobutton?

I am missing the context here. Is it in regard to the screenshot above? What is to be chosen?

**Comment by VicDeo** [comment]
@felixheidecke the context is providing two versions for update: minor and major,
see #387 for more details

**Comment by felixheidecke** [comment]
The dropdowns in the screenshot in the very first post look good. Don't they suffice?

File: lib/Command/InstallApp.php [hunk 2]
```diff
@@ -85,12 +93,14 @@ protected function execute(InputInterface $input, OutputInterface $output) {
...
+						try {
+							$this->checkVersion($currentVersion, $packageVersion);
 							$output->writeln("$appId: Installing new version from $localPackage");
 							$this->marketService->updatePackage($localPackage);
 							$output->writeln("$appId: App updated.");
-						} else {
-							$output->writeln("$appId: $localPackage has the same or older version of the app");
+						} catch (\DomainException $e) {
+							$errorMessage = $e->getMessage();
+							$output->writeln("$appId: $localPackage $errorMessage");
```
Comment (PVince81) [inline lib/Command/InstallApp.php]: no stderr output possible I guess ?

File: lib/HttpService.php [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+			}
+		}
+
+		$this->checkInternetConnection();
+		$apiKey = $this->getApiKey();
+		$endpointUrl = $this->getAbsoluteUrl($uri);
+		$response = $this->httpGet($endpointUrl, [], $apiKey);
+		$data = $response->getBody();
+		if ($this->cacheFactory->isAvailable()) {
+			// cache if for a day - TODO: evaluate the response header
```
Comment (PVince81) [inline lib/HttpService.php]: unimportant todo ?

File: tests/unit/HttpServiceTest.php [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+			$this->config,
+			$this->cacheFactory,
+			$this->l10n
+		);
+	}
+
+	/**
+	 * @dataProvider testCheckInternetConnectionDataProvider
+	 */
+	public function textCheckInternetConnection($connectionStatus, $expectedExceptionClass) {
```
Comment (PVince81) [inline tests/unit/HttpServiceTest.php]: text ? this test likely did not run

File: tests/unit/HttpServiceTest.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+	public function testCheckInternetConnectionDataProvider() {
+		return [
+			[true, ''],
+			[false, AppManagerException::class]
+		];
+	}
+
+	public function testGetApps() {
+		$expectedApps = [];
```
Comment (PVince81) [inline tests/unit/HttpServiceTest.php]: no expected apps ?

File: lib/HttpService.php (hunk 3 above)
Comment (VicDeo) [reply lib/HttpService.php]: not mine, I just copied it while extracting methods from MarketService

File: tests/unit/HttpServiceTest.php (hunk 5 above)
Comment (VicDeo) [reply tests/unit/HttpServiceTest.php]: yes. the amount makes no difference from the point of coverage. :)

File: lib/Command/InstallApp.php (hunk 2 above)
Comment (VicDeo) [reply lib/Command/InstallApp.php]: Possible but we don't use stderr for other errors so it's better leave it as is.
As for markup - this is not a fatal error, so it merely could be wrapped with `<error>`.

File: tests/unit/HttpServiceTest.php (hunk 4 above)
Comment (VicDeo) [reply tests/unit/HttpServiceTest.php]: Oops. Fixed.

Comment (VicDeo) [reply tests/unit/HttpServiceTest.php]: 0.22% more to the coverage :)

File: lib/MarketService.php (hunk 1 above)
Comment (VicDeo) [reply lib/MarketService.php]: changed to have only an exact version as an optional parameter, no  more `isMajor`

**Comment by VicDeo** [comment]
@PVince81 answered everywhere

**Comment by davitol** [comment]
In this scenario:

screen shot 2018-11-22 at 10 19 00

Do we feel OK with the text shown? Is it clear enough for an admin?

```password_policy: No compatible version for password_policy```


screen shot 2018-11-22 at 10 19 55

(cc @pmaier1 )

**Comment by pmaier1** [comment]
> Do we feel OK with the text shown? Is it clear enough for an admin?

When exactly is this message shown? Only when there's a requirement for `--major` to be able to upgrade the app or also when there is no later version available at all or even when you are on the latest version already? Anything else?

**Comment by PVince81** [comment]
let's clarify UX and messages in https://github.com/owncloud/market/issues/416

**Pull request merged**